writer.go

```diff
@@ -0,0 +1,383 @@
+// correctly.
+type Writer struct {
+	// Header is written to the file when [Writer.Close] is called.
+	Header
+
+	// tmp is the temporary file where chunks will be written.
+	tmp *os.File
+
+	// hasData is true if data has been written to the chunk buffer but hasn't
+	// been finalized and written to tmp. We need this because we can't simply
+	// call z.Flush and check chunkBuf.Len due to the fact that flate.Writer
+	// will write sync markers on every call to Flush even if no data has been
+	// written.
+	hasData bool
+
+	// chunkBuf is the current compressed chunk.
+	chunkBuf *bytes.Buffer
+
+	// z is the compression writer used to write the current compressed chunk
+	// to chunkBuf.
+	z *flate.Writer
+
+	// zMW is an io.MultiWriter that writes to both z (flate.Writer) and
+	// digest (hash.Hash32).
+	zMW io.Writer
+
+	// w is the io.Writer for the final destination for the compressed file.
+	w io.Writer
+
+	// digest is the CRC-32 digest (IEEE polynomial).
+	// See RFC-1952 Section 2.3.1.
+	digest hash.Hash32
+
+	// isize is the total size of the uncompressed input.
+	isize int64
+
+	// level is the compression level being used.
+	level int
+
+	// closed indicates the writer has been closed.
+	closed bool
+}
+
+// NewWriter initializes a new dictzip Writer with the default compression
+// level and chunk size.
+func NewWriter(w io.Writer) (*Writer, error) {
+	return NewWriterLevel(w, DefaultCompression, DefaultChunkSize)
+}
+
+// NewWriterLevel initializes a new dictzip Writer with a compression level and
+// chunk size.
+func NewWriterLevel(w io.Writer, level, chunkSize int) (*Writer, error) {
+	tmp, err := os.CreateTemp("", "dictzip.*")
+	if err != nil {
+		return nil, fmt.Errorf("creating temp file: %w", err)
+	}
+
+	var buf bytes.Buffer
+	fw, err := flate.NewWriter(&buf, level)
+	if err != nil {
+		return nil, fmt.Errorf("initializing deflate writer: %w", err)
+	}
+
+	digest := crc32.NewIEEE()
+	z := Writer{
+		tmp:      tmp,
+		hasData:  false,
+		chunkBuf: &buf,
+		z:        fw,
+		zMW:      io.MultiWriter(fw, digest),
+		w:        w,
+		digest:   digest,
+		level:    level,
+	}
+	z.chunkSize = chunkSize
+
+	return &z, nil
+}
+
+func (z *Writer) Write(p []byte) (int, error) {
+	if z.closed {
+		return 0, fmt.Errorf("%w: Write called on closed writer", errDictzip)
+	}
+
+	// Write chunks to z.z, resetting the Writer, and flushing chunks to the
+	// z.tmp as necessary.
+	var i int
+	for i < len(p) {
+		// Get the end index by adding the chunk size minus any already written
+		// part of the current chunk.
+		j := i + z.chunkSize - int(z.isize%int64(z.chunkSize))
+		if j > len(p) {
+			j = len(p)
+		}
+
+		// NOTE: Write to the multi-writer to update the CRC digest.
+		n, err := z.zMW.Write(p[i:j])
+		z.isize += int64(n)
+		i += n
+		if err != nil {
+			return i, fmt.Errorf("compressing: %w", err)
+		}
+		if n > 0 {
+			z.hasData = true
+		}
+
+		if z.isize%int64(z.chunkSize) == 0 {
```

Comment:
The check of whether a full chunk has been written is being done in lines 137-140. Consider saving a bool value and using it here as that might be clearer.